### PR TITLE
Clustering

### PR DESCRIPTION
**Pull request by Chocomunk**
Making this pull request so I can make comments on your code.

### CODE REVIEW

File: util/clustering/kmeans_clustering.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+from tqdm import tqdm
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Clustering')
+    if 'SM_OUTPUT_DATA_DIR' in os.environ: 
+
+        parser.add_argument('--output-data-dir', type=str, default=os.environ['SM_OUTPUT_DATA_DIR'])
+        #parser.add_argument('--model-dir', type=str, default=os.environ['SM_MODEL_DIR'])
+        parser.add_argument('--input-data-dir', type=str, default=os.environ['SM_MODEL_DIR'])
```
Comment (Chocomunk) [inline util/clustering/kmeans_clustering.py]: 'SM_MODEL_DIR' refers to literally where we should save the model to. This should be assigned to the '--model-dir' parameter

File: util/clustering/kmeans_clustering.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+import matplotlib.pyplot as plt
+from PIL import Image
+from tqdm import tqdm
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Clustering')
+    if 'SM_OUTPUT_DATA_DIR' in os.environ: 
+
+        parser.add_argument('--output-data-dir', type=str, default=os.environ['SM_OUTPUT_DATA_DIR'])
```
Comment (Chocomunk) [inline util/clustering/kmeans_clustering.py]: 'SM_OUTPUT_DATA_DIR' is completely managed by SageMaker backend, we don't want that (we want to control where our files are going). For now, set this argument to have no default (we can discuss this later when we put it on SageMaker).

File: util/clustering/kmeans_clustering.py (hunk 1 above)
Comment (Chocomunk) [reply util/clustering/kmeans_clustering.py]: Same thing as above, no default for now

File: util/clustering/kmeans_clustering.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        #parser.add_argument('--model-dir', type=str, default=os.environ['SM_MODEL_DIR'])
+        parser.add_argument('--input-data-dir', type=str, default=os.environ['SM_MODEL_DIR'])
+    else: 
+        parser.add_argument('--output-data-dir', type=str)
+        parser.add_argument('--input-data-dir', type=str)   
+    
+    parser.add_argument('--analyze-clusters', type=int, default=3)
```
Comment (Chocomunk) [inline util/clustering/kmeans_clustering.py]: Bad parameter name. This parameter represents the number of clusters, so name it something related (e.g. '--num-clusters').

File: util/clustering/kmeans_clustering.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    
+    parser.add_argument('--analyze-clusters', type=int, default=3)   
+
+    args = parser.parse_args()
+
+    
+    file_dir = os.getcwd()
+    # # change the working directory to the path where the images are located
+    output_path = args.input_data_dir
+    os.chdir(output_path)
```
Comment (Chocomunk) [inline util/clustering/kmeans_clustering.py]: This is bad practice. Don't chdir into the directories where we want to do file management, just incorporate the path int the name when reading/writing to files.

File: util/clustering/kmeans_clustering.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+    # # this list holds all the image filename
+    # detection_train_images = [f.key.split(detection_train_images_path + "/")[1] for f in s3_bucket.objects.filter(Prefix=detection_train_images_path).all()]
+    print(str(len(detection_train_images)) + ' images collected')
+
+        
+    model = VGG16()
+    model = Model(inputs = model.inputs, outputs = model.layers[-2].output)
+
+    def extract_features(file, model):
```
Comment (Chocomunk) [inline util/clustering/kmeans_clustering.py]: This does not need to be a private inner-function. Just define it as a normal function outside of main()

Comment (Chocomunk) [reply util/clustering/kmeans_clustering.py]: in-fact it really should not be an inner function. The variable names in this scope are conflicting and can cause logical errors for our variable references.

File: util/clustering/kmeans_clustering.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+        img = np.array(img) 
+        # reshape the data for the model reshape(num_of_samples, dim 1, dim 2, channels)
+        reshaped_img = img.reshape(1,224,224,3) 
+        # prepare image for model
+        imgx = preprocess_input(reshaped_img)
+        # get the feature vector
+        features = model.predict(imgx, use_multiprocessing=True)
+        return features
+    
+    data = {}
```
Comment (Chocomunk) [inline util/clustering/kmeans_clustering.py]: Bad variable name. 'data' is actually storing image features, not input data or another generic form

File: util/clustering/kmeans_clustering.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+    args = parser.parse_args()
+
+    
+    file_dir = os.getcwd()
+    # # change the working directory to the path where the images are located
+    output_path = args.input_data_dir
+    os.chdir(output_path)
+
+    detection_train_images = []
```
Comment (Chocomunk) [inline util/clustering/kmeans_clustering.py]: The variable naming isn't terrible, but it will get confusing if we want to use it on other image sets. A better name would be simple 'images' since there are no other images in the entire script.

Comment (Chocomunk) [reply util/clustering/kmeans_clustering.py]: Sorry I didn't finish typing the name: 'image_fname'

File: util/clustering/kmeans_clustering.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+        data[detection_train_images[i]] = feat
+        
+        # if something fails, save the extracted features as a pickle file (optional)
+        # except:
+        #     with open(p,'wb') as file:
+        #         pickle.dump(data,file)
+            
+    
+    # get a list of the filenames
+    filenames = np.array(list(data.keys()))
```
Comment (Chocomunk) [inline util/clustering/kmeans_clustering.py]: This is turning a list of strings into a numpy array for no reason (that I can see). This is unnecessary and doesn't really make sense, just leave it as  a python list.

File: util/clustering/kmeans_clustering.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+
+    # get a list of just the features
+    feat = np.array(list(data.values()))
+
+    # reshape so that there are 210 samples of 4096 vectors
+    feat = feat.reshape(-1,4096)
```
Comment (Chocomunk) [inline util/clustering/kmeans_clustering.py:129]: Is 4096 the output dim of VGNet? What are the 210 samples, I thought there were 1000?

The comment doesn't really describe what is happening here. Only talk about what the code is doing generically (e.g. "split into (samples, features)"), not for a specific run on specific inputs (although I still don't know where 210 came from).

File: util/clustering/kmeans_clustering.py [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+    # # Plot sse against k
+    # print('------------ plotting sse against k ------------')
+    # plt.figure(figsize=(6, 6))
+    # plt.plot(list_k, sse)
+    # plt.xlabel(r'Number of clusters *k*')
+    # plt.ylabel('Sum of squared distance')
+    # plt.show()
+    # plt.savefig(args.output_data_dir + '/cluster_analysis.png')
+
+    # num_clusters = best_k
```
Comment (Chocomunk) [inline util/clustering/kmeans_clustering.py]: If the comments are left here for testing purposes, then can you move them into separate functions? They make the code hard to read.

If they aren't for testing or any other specific purpose (e.g. "idk if I need this later" is a good specific purpose) then just remove them.

File: util/clustering/kmeans_clustering.py [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+    print('------------ creating label files ------------')
+
+
+    for cluster in groups.keys(): 
+        with open(f"labels/label_{cluster}.txt", "w") as output:
+            for file in groups[cluster]:
+                output.write('%s\n' % file)
+
+    # function that lets you view a cluster (based on identifier)        
+    def view_cluster(cluster):
```
Comment (Chocomunk) [inline util/clustering/kmeans_clustering.py]: Again, there isn't a reason for this to be an inner-function. Should move this outside of main()

File: util/clustering/kmeans_clustering.py [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+    # reshape so that there are 210 samples of 4096 vectors
+    feat = feat.reshape(-1,4096)
+
+
+    # # reduce the amount of dimensions in the feature vector
+    print('------------ feature reduction ------------')
+    pca = PCA(n_components=100, random_state=22)
+    x = pca.fit_transform(feat)
+
+    os.chdir(file_dir)
```
Comment (Chocomunk) [inline util/clustering/kmeans_clustering.py]: Same thing as above, don't chdir just use the path

File: util/clustering/kmeans_clustering.py (hunk 11 above)
Comment (Chocomunk) [reply util/clustering/kmeans_clustering.py]: Also, like above, don't chdir

**Review by Chocomunk — COMMENTED** [review]
Here are things I noticed while reading through the code. Please fix them and then let me review it again. 

Don't merge when you are done, I'll merge once everything is done being reviewed.

File: util/clustering/kmeans_clustering.py (hunk 8 above)
Comment (matthewGonzalgo) [reply util/clustering/kmeans_clustering.py]: I think it's just because numpy arrays have better performance?

File: util/clustering/kmeans_clustering.py (hunk 9 above)
Comment (matthewGonzalgo) [reply util/clustering/kmeans_clustering.py:129]: I believe 4096 is the output dim of VGNet but not entirely sure. 
Sorry I should have removed the line mentioning 210 samples. That is the size of the dataset they used in the blog post

Comment (matthewGonzalgo) [reply util/clustering/kmeans_clustering.py:129]: Here's the VGGNet architecture:
